src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,362 @@
+        {
+            // From http://dev.w3.org/html5/spec/Overview.html#elements-0
+            private static readonly HashSet<string> VoidElements = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+            {
+                "area",
+                "base",
+                "br",
+                "col",
+                "command",
+                "embed",
+                "hr",
+                "img",
+                "input",
+                "keygen",
+                "link",
+                "meta",
+                "param",
+                "source",
+                "track",
+                "wbr"
+            };
+
+            private readonly Stack<TagBlockTracker> _startTagTracker;
+
+            public Rewriter()
+            {
+                _startTagTracker = new Stack<TagBlockTracker>();
+            }
+
+            private TagBlockTracker CurrentTracker => _startTagTracker.Count > 0 ? _startTagTracker.Peek() : null;
+
+            private string CurrentStartTagName => CurrentTracker?.TagName;
+
+            public override SyntaxNode Visit(SyntaxNode node)
+            {
+                if (node != null)
+                {
+                    node = RewriteNode(node);
+                }
+
+                return base.Visit(node);
+            }
+
+            private SyntaxNode RewriteNode(SyntaxNode node)
+            {
+                if (node.IsToken ||
+                    node is MarkupElementSyntax ||
+                    node.Parent is MarkupElementSyntax)
+                {
+                    // This is either a token or we have already rewritten this node.
+                    return node;
+                }
+
+                var currentIndex = 0;
+                _startTagTracker.Clear();
+                var children = node.ChildNodes();
+                while (currentIndex < children.Count)
+                {
+                    var child = (RazorSyntaxNode)children[currentIndex];
+                    if (!(child is MarkupTagBlockSyntax tagBlock) ||
+                        tagBlock.Parent is MarkupElementSyntax)
+                    {
+                        // Not a node that we want to rewrite.
+                        currentIndex++;
+                        continue;
+                    }
+
+                    var tagName = tagBlock.GetTagName();
+                    if (string.IsNullOrWhiteSpace(tagName) ||
+                        IsVoidElement(tagBlock) ||
+                        IsSelfClosing(tagBlock))
+                    {
+                        // Don't want to track incomplete, void or self-closing tags.
+                        // Simply wrap it in a block with no body or end tag.
+                        node = RewriteNodeCore(node, startTag: tagBlock, tagChildren: Enumerable.Empty<RazorSyntaxNode>(), endTag: null);
+
+                        // Since we rewrote 'node', it's children are different. Update our collection.
+                        children = node.ChildNodes();
```

Comment:
How does this work? Should this restart from the beginning since the children are different or continue from the next index because this node is already observed?

Reply:
It won't restart. It just continues from where it left off. I still have to update the collection because rewriting node changes the children.

Reply:
What if instead you created a new list of children and then rewrote it once.

So like:
```
A
  B 
  C 
  D
```

Iterate through the children of `A` - `(B, C, D)` and rewrite just those nodes as needed so you end up with `(B, C!, D)` (assume C was rewritten). Then rewrite `A` once to replace its children. That's how I would expect something like this to work.

Reply:
The problem is that when I rewrite (B, C, D) => (B, C!, D), it actually becomes (B, C!, D!) because everything after a rewritten node is different because red nodes are created on demand

Reply:
Can you explain this in more detail? I'm not sure that I get it.

Reply:
Sorry I missed this part "created a **new** list of children". What I said doesn't apply in that case. I'll give this a shot